### DISCUSSION TITLE
[feature] Computed Configuration files

### DISCUSSION THREAD

**Discussion by JeppeKlitgaard** [espanso/espanso#1218] [Feature requests and Ideas]
Hey @federico-terzi!

I wanted to expand on the discussion we were having over in #632 and decided that a separate issue would probably be appropriate at this point, though the solution I will propose here would also fix #632 along with a number of other issues relating to configuration flexibility.

I started playing around with `jsonnet` which is essentially a small programming language targeted at writing configuration files. It can load JSON, arbitrary strings, and very soon also YAML (see: https://github.com/google/go-jsonnet/pull/339). 

Similarly, it can also output JSON and YAML (the YAML output is more 'yaml'-like than just using the JSON directly in YAML).

It is flexible and allows features like:
- Local variables
- Imports
- Functions

I think this could be the way to go for packages as well, such that packages would effectively have to expose a simple jsonnet function that the user would import in their local configuration in order to smooth out any edges that they might have with their individual set (like a different choice of triggers).

I have a quick proof of concept for global triggers, but obviously the capabilities of jsonnet go far beyond this:

```jsonnet
# file: base.libsonnet
local addPrePost(trigger, pre, post) =
    pre + trigger + post
;

{
    PRE: ":",
    POST: " ",

    processRawMatches(rawMatches, pre, post): [
        {
            [if "triggers" in match then "triggers"]:
                [
                    addPrePost(trigger, pre, post)

                    for trigger in match["triggers"]
                ],

            [if "trigger" in match then "trigger"]: addPrePost(match["trigger"], pre, post),

            replace: match["replace"],
        },
        for match in rawMatches
    ]
}
```

```jsonnet
# file: fractions.jsonnet
local g = import 'base.libsonnet';

local rawMatches = [
  # 2
  {
    triggers: [
      'half',
      '1/2',
    ],
    replace: '½',
  },

  # 3
  {
    triggers: [
      'third',
      '1/3',
    ],
    replace: '⅓',
  },
  {
    trigger: '2/3',
    replace: '⅔',
  },
 # And lots more in actuality, cut short of brevity
];

std.manifestYamlDoc(
  {
    name: 'fractions',
    parent: 'default',

    matches: g.processRawMatches(rawMatches, g.PRE, g.POST),
  }
)
```

I can then produce files for `espanso` using:
```sh
jsonnet -S fractions.jsonnet
```

Which outputs:
```yaml
"matches":
- "replace": "½"
  "triggers":
  - ":half "
  - ":1/2 "
- "replace": "⅓"
  "triggers":
  - ":third "
  - ":1/3 "
- "replace": "⅔"
  "trigger": ":2/3 "
"name": "fractions"
"parent": "default"
```

---

This works well even without a native integrating with espanso, but using https://github.com/CertainLach/jrsonnet it could be even better.

Since `jsonnet` can output to yaml, you'd just have to run `jsonnet` on `.jsonnet` files in `.../user/`, and these would be able to live peacefully alongside existing yaml files.

---

Apologies for the long issue. I will continue to explore jsonnet used alongside `espanso`, which I think could be a super powerful combination!

**Comment by federico-terzi** [comment]
Hey @JeppeKlitgaard,

Very interesting proposal! Thank you. As I said in #632, I think integrating this feature within espanso itself would make it overly complex, but I can definitely see how this could be useful in various situations.
I think this could be a great opportunity to make an external utility that could be "plugged in" with espanso to generate those configs if needed.

What do you think?

Cheers :)

**Comment by JeppeKlitgaard** [comment]
@federico-terzi I definitely agree that implementing something like this from the ground up would be way too complex, but when I found jsonnet, I thought it might just be able to work.

I am mostly comfortable in Python, but I trust that the Rust bindings would be similarly simple. Importing a jsonnet file in python is as simple as:

```python
import json
import _jsonnet

jsonnet_file = "jsonnet/bar.jsonnet"
data = json.loads(_jsonnet.evaluate_file(jsonnet_file))
print("{str} => {num}".format(**data))
```

If you prefer YAML, you'd just use a yaml loader instead of json and change the jsonnet file to output yaml.

(taken from: https://stackoverflow.com/questions/49949418/how-to-correctly-call-jsonnet-with-imports-from-python)

I think you're right that this definitely does clash with the goals of a GUI and as for the dynamic variables I am not entirely sure I understand why this shouldn't work with jsonnet. Is the current implementation re-reading the yaml config every time a script is executed? A solution in that case would be to just read and compile the jsonnet file once and store the contents as a string. That way there wouldn't be a need to recompile (very possible I am missing something here)

Possible solution to GUI:
- Since jsonnet support does not replace yaml support, the GUI would simply build/read yaml files. If the advanced features of jsonnet are needed, one would have to resort to using an editor with the jsonnet files.

That being said, it could work almost as well as a separate tool, though requiring slightly more tinkering from users. The disadvantage of doing it this way would be:
- Not possible to use jsonnet to enrich the package infrastructure
- No automatic update of changes in jsonnet files
- Users would have to set up the tooling themselves

If I get around to it (exam season 😢) I'll give a quick demonstration of my current setup, which I hacked together in a few hours today. It features:
- Programmatically generated unicode replacements (replacements of `:BBBa` -> 𝕒 for a-z, A-Z in a few lines of code)
- Compile jsonnet to yaml with a `:compileJsonnet` match
- Programmatically adding `propagate_case` to existing matches
- Flexible addition of pre/post triggers.

(currently you would be able to find a very rough version of this in https://github.com/JeppeKlitgaard/dotfiles/tree/master/.config/espanso, but I will improve upon this in the future.

For example, the Blackboard Bold file is not done yet, though the unicode ranges are found)

---

The most powerful use-case for jsonnet would probably be to make packages more flexible. I appreciate that the refactor is your first priority at the moment though!

**Comment by federico-terzi** [comment]
@JeppeKlitgaard Thank you for sharing your setup and thoughts, those are indeed good points!

Let's keep this issue open for discussion, keeping in mind that it has low-priority at the moment :)
